### ISSUE TITLE
Serialization bug in new codegenerator for types deriving from Dictionary<TKey, TValue>

### ISSUE THREAD

**Issue by rmja** [dotnet/orleans#5029]
If a grain interface method takes as parameter a custom dictionary which inherits from `Dictionary<TKey, TValue>` then the dictionary is not populated on the receiving side. The error only exist when the new code generator is used, it works correctly with the old one.

There is a reproducing project here: https://github.com/rmja/OrleansIssueDictionary
The project currently references the new generator. Changing back to the old one makes the test pass.

**Comment by rmja** [comment]
I could probably give this one a crack during the hacktoberfest :-)

**Comment by ReubenBond** [comment]
@rmja the issue here is that the code generator is unable to find the private fields in that type because the version of mscorlib.dll which is being referenced during compilation time is a reference assembly which has had all private members stripped.

I'm open to suggestions on how to remedy it.

One approach is to use a hand-crafted serializer for the base class (`Dictionary<K, V>`), but that would require larger change to the serialization system. My side-project serializer implements that strategy.

Currently the plan is to warn users about this issue during build time.

**Comment by rmja** [comment]
Holy moly, I thought it was something simple. What would be the recommended way to to mitigate it then? I guess I could implement `IDictionary<K,V>` and have the actual `Dictionary<K,V>` instance as a member in that implementation - would that work?

**Comment by rmja** [comment]
Year, I just tested that - so instead of:

```
public class Hash : Dictionary<Key, Value>
{
}
```

One can:
```
public class Hash : IDictionary<Key, Value>
{
    private readonly Dictionary<Key, Value> _dictionary = new Dictionary<Key, Value>();

    // Implement IDictionary through _dictionary here...
}
```

**Comment by ReubenBond** [comment]
@rmja yep that works since it will use the in-build `Dictionary<K, V>` serializer.

This is a significant detractor the the new code generator at this time. Something which will need more thought. Open to input, as always.

**Comment by denisivan0v** [comment]
I've faced with the same issue, workaround with changing from inheritance to composition works.
However, it doesn't look like a proper solution.

**Comment by ReubenBond** [comment]
Fixed in 4.0.0-preview1
